Add Getting Started > Basics documentation (#4914)
openshift/odo: website/docs/getting-started/basics.md

---
# odo concepts

odo abstracts Kubernetes concepts into a developer friendly terminology. In this document, we will take a look at the terminology used by odo.

#### Application
  An application in odo is a classic application developed with a [cloud-native approach](https://www.redhat.com/en/topics/cloud-native-apps) that is used to perform a particular task.
  
  Examples of applications: Online Video Streaming, Hotel Reservation System, Online Shopping.

#### Component

>   In the cloud-native architecture, an application is a collection of small, independent, and loosely coupled components; and an odo component is one of these components.

Reviewer: ```suggestion
  In the cloud-native architecture, an application is a collection of small, independent, and loosely coupled components; an odo component is one of these components.
```

  Examples of components: Warehouse API Backend, Inventory API, Web Frontend, Payment Backend.

#### Project

>   A project helps achieve multi-tenancy: several applications can be run in the same cluster by different teams.

Reviewer: ```suggestion
  A project helps achieve multi-tenancy: several applications can be run in the same cluster by different teams in different projects.
```

#### Context
  A context is the directory on the system that contains the source code, tests, libraries and odo specific config files for a single component.

#### URL

>   A URL exposes a component to be accessed outside the cluster.

Reviewer: ```suggestion
  A URL exposes a component to be accessed from outside the cluster.
```

#### Storage

>   A storage is persistent storage in the cluster: it persists the data across restarts and rebuilds of a component.

Reviewer: ```suggestion
  A storage is a persistent storage in the cluster: it persists the data across restarts and rebuilds of a component.
```

---
# odo concepts

odo abstracts Kubernetes concepts into a developer focused terminology. In this document, we will take a look at each of these concepts.

### Concepts

* **Component** 
  
  A component is like a microservice. Each component can be run and deployed separately. Examples of components: Warehouse API Backend, Inventory API, Web Frontend, Payment Backend

* **Application**
  
  An application is a group of one or more components that work individually to build the entire application. Examples of applications: Online Video Streaming, Hotel Reservation System, Online Booking

* **Project**
  

>   A project is a separate single unit that provides a scope for names and helps in dividing cluster resources between users. A resource name must be unique within the project but not across multiple projects.

Reviewer 2: Wow! That sounds like a textbook definition. Where did you get that from?
```suggestion
  A project helps achieve multi-tenancy. Resource names must be unique within the project but not across multiple projects.
```

* **Context**
  

>   A context is the directory where the source code, tests, libraries and odo specific config files for your component resides. A single context can only contain one component.

Reviewer 2: IMO, we should point that context "resides" on the user's system.
```suggestion
  A context is the directory on your system where the source code, tests, libraries and config files for a component reside. Loosely speaking, it is a locally cloned git repository and odo specific config files (created by `odo create`). A single context can only contain one component.
```

* **URL**
  
  A URL exposes your component to be accessed outside the cluster.

* **Storage**
  

>   A storage is a way to claim persistent storage in the cluster. A storage can persist data across restarts and rebuilds of a component.

Reviewer 2: ```suggestion
  A storage is persistent storage in the cluster. It helps persist data across restarts and rebuilds of a component.
```

* **Service**
  

>   A service is an external application that a component can connect to or depend on to gain some functionality. Example of services: MySQL, Redis

Reviewer 2: ```suggestion
  A service is an external application that a component can connect to or depend on to gain some functionality. Example of services: MySQL, Redis.
```

* **Devfile**
  
  A portable YAML file responsible for your entire reproducible development environment. See [Devfile](../architecture/devfile.md) to know more about devfile.
  

### odo services vs. Kubernetes services

> Service in terms of odo is an external application that provides an additional functionality, odo service can only be created with [Operators](https://operatorframework.io/what/).

Reviewer 2: ```suggestion
Service, in terms of odo, is an external application that provides some extra functionality. odo services can only be created with [Operators](https://operatorframework.io/what/).
```

> Service in terms of Kubernetes is an abstract way of exposing a microservice as a set of pods. Learn more about [Kubernetes Service](https://kubernetes.io/docs/concepts/services-networking/service/).

Reviewer 2: ```suggestion
Service, in terms of Kubernetes, is an abstract way of exposing a microservice. Learn more about [Kubernetes Service](https://kubernetes.io/docs/concepts/services-networking/service/).
```

> Deploying a service created with odo will create a set of resources in the Kubernetes cluster to establish a successful connection between the odo component and the external application.

Reviewer 2: I strongly feel like we should have this differentiation in the Command Reference section where we talk about `odo service` commands.

Without creating a link, the below statement is incorrect:
> establish a successful connection between the odo component and the external application.
```suggestion
Deploying a service created with odo will create a set of resources in the Kubernetes cluster which can then be used from within an odo component after performing link operation.
```

### Concepts

* Component

>   A component is a microservice. Each component can be run and deployed separately. Examples of components: Warehouse API Backend, Inventory API, Web Frontend, Payment Backend

Author: Yeah, my bad, I was too focused on the quoted part. I'll fix that.

>   > A component in odo is an equivalent of a deployment or deploymentconfig in Kubernetes. Creating a new component in odo will create a deployment in Kubernetes or for git and binary based component creation, it will create deploymentconfig.

Reviewer 2: 1. Why is this quoted with a `>`? Same question for everywhere else in the doc.
2. IMO, odo component is not an equivalent of Deployment or DeploymentConfig. By default, odo creates a bunch of Kubernetes resources for a component:
    - Deployment (which creates a ReplicaSet and Pods)
    - Service (a Kubernetes Service; not odo service)
    - Storage (emptyDir or PVC depending on user's preference setting)
    - probably more that I might have missed.
3. We don't need to go into this level of detail for git/binary components. Also, it doesn't create just DC but other things that I mentioned above.

Reviewer: I understand that the quoted part explain how the concepts are implemented in Kubernetes.

But it is probably not a good place to talk about these implementations. This concept page should help the reader make a mental representation of the objects handled by odo, without other distraction.

Author: >Why is this quoted with a >? Same question for everywhere else in the doc.

1. I figured we would put the Kubernetes part in a quoted text like [Philippe has done in his blog](https://feloy.medium.com/odo-a-tool-to-simplify-development-on-kubernetes-ce0df34313b7).
I can add a note about it in the beginning.

>IMO, odo component is not an equivalent of Deployment or DeploymentConfig. By default, odo creates a bunch of Kubernetes resources for a component:

    Deployment (which creates a ReplicaSet and Pods)
    Service (a Kubernetes Service; not odo service)
    Storage (emptyDir or PVC depending on user's preference setting)
    probably more that I might have missed.

2. Okay, I can change that.

>We don't need to go into this level of detail for git/binary components. Also, it doesn't create just DC but other things that I mentioned above.

3. Noted.

Author: >But it is probably not a good place to talk about these implementations. This concept page should help the reader make a mental representation of the objects handled by odo, without other distraction.

Okay okay. I will remove the quoted texts. @feloy, let me know if you have something specific in mind about formatting this.

Author: > >I understand that the quoted part explain how the concepts are implemented in Kubernetes.
> >
> > But it is probably not a good place to talk about these implementations. This concept page should help the reader make a mental representation of the objects handled by odo, without other distraction.
> 
> Okay okay. I will remove the quoted texts. @feloy, let me know if you have something specific in mind about formatting this.

On a second thought, we mention early on in this page that we will be talking about these concepts and their Kubernetes equivalent, so it does make sense to me to put them here.

Reviewer 2: > I figured we would put the Kubernetes part in a quoted text like [Philippe has done in his blog](https://feloy.medium.com/odo-a-tool-to-simplify-development-on-kubernetes-ce0df34313b7).
>        I can add a note about it in the beginning.

We don't want to introduce k8s lingo here.

> On a second thought, we mention early on in this page that we will be talking about these concepts and their Kubernetes equivalent, so it does make sense to me to put them here.

My bad. We should not talk about k8s equivalents in this doc. Can you remove that statement as well?

Author: Okay. Where else do we mention it then? This seemed like the right place to me to explain about the K8s equivalent of these concepts.

Reviewer: We could limit talking about Kubernetes in the Command Reference and Architecture pages

Author: But I do think that talking about the K8s equivalent here will help get a better understanding of odo in general. We do not need to talk about the deep implementation level details, just the high-level ones to make a relation between odo terminology and K8s terminology.

I understand that it might defeat the purpose of abstracting K8s out of the docs, but if I was reading this doc as odo user, having this relation would help me better understand the docs. Or I guess we can make a better effort at making these concepts better understandable.

Reviewer 2: > I understand that it might defeat the purpose of abstracting K8s out of the docs,

This is why we don't want to introduce k8s lingo so early.

> but if I was reading this doc as odo user, having this relation would help me better understand the docs

We do not want application developers (our target audience) to deal with k8s lingo. It's for application architects and whoever _already_ has knowledge of k8s. Majority of odo users are not expected to have knowledge of k8s.

> Or I guess we can make a better effort at making these concepts better understandable.

And that we'll be doing in Command Reference and Architecture sections. Maybe at some places in tutorials where it really makes sense. But Getting Started is definitely not the place to introduce k8s lingo.

* Application

>   An application consists of multiple components which may span over multiple projects, and work individually to build the entire application. Examples of applications: e-Shop, Hotel Reservation System, Online Booking

Reviewer 2: > An application consists of multiple components which may span over multiple projects

This is not true for odo. odo doesn't cover things across multiple namespaces.

Author: I meant that an application can be accessed over multiple namespaces. But I see how it might have come off to mean that a user can access the same component over multiple namespaces, is that what you meant?

Anyway, I'll try to fix that.

Reviewer 2: What I understood is that the components can be distributed over different namespaces. Which is not true for odo. At least not yet.

  > An application can be considered as an equivalent of labels in Kubernetes that help in grouping a set of resources.
* Project

>   A project is your source code, tests, and libraries organized in a separate single unit.

Reviewer 2: This sounds more like a definition of what a component directory (a.ka. context) is on the dev's local system.

Reviewer: Without naming the underlying Kubernetes `namespace` resource, you can explain the motivations of using separate namespaces:
- a scope for names
- divide cluster resources between users (resources quotas, RBAC are based on namespaces)

https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/

Author: Yeah, I actually picked that up from one of the existing docs. I'll try to fix that.

  > A project in odo is an equivalent of a namespace in Kubernetes. Creating a project in odo will create a namespace in Kubernetes with the same name.
* Context
  A context is the directory where the source code, tests, and libraries for your component resides. A single context can only contain a single component.
* URL
  A URL exposes your component to be accessed outside the cluster.

>   > A URL is an equivalent of a service in Kubernetes.

Reviewer 2: NO! It's not an equivalent of a Service in k8s! It's equivalent to Ingresses in k8s and Routes in OCP.

Author: :grimacing: Okay. Changing that.

Author: The OCP part was just for my fyi, right? Do we want to mention it in the docs?

Reviewer 2: Yeah. Even the Ingress thing was for reference. We need to explain URL without talking about Ingresses or Routes.

* Storage

>   A storage volume is [PVC](https://kubernetes.io/docs/concepts/storage/volumes/#persistentvolumeclaim) which is a way for you to "claim" persistent storage without knowing the details of the environment. Storage volume can persist data across restarts and rebuilds of a component.

Reviewer 2: Using k8s terminology to explain odo concept is not a good idea. It just means we fail to explain (or even understand ourselves!) the abstractions we created to make k8s terminology not required for the user.

Author: Again, picked it up from existing docs, and it seemed to make sense to me, but I get your point.

  > A storage volume is an equivalent of PVC in Kubernetes. Creating a storage with odo will create a PVC resource in Kubernetes with the same name.
* Service

>   A service is another microservice, or a Kubernetes Custom Resource that your component connects to or depends on. Example of services: MariaDB, MySQL

Author: How would you define an operator-backed service? If I use the term, I will still need to explain what a service is. How do I define a service?
By "microservice" I meant component. Ah, I confused service with link :facepalm:

>   A service is another microservice, or a Kubernetes Custom Resource that your component connects to or depends on. Example of services: MariaDB, MySQL

Reviewer 2: > How would you define an operator-backed service? If I use the term, I will still need to explain what a service is. How do I define a service?

A loose definition of service is a long-running daemon (think a server process), and listening for incoming requests based on which it would do certain actions. But we don't need to explain this in our docs. It's a definition that almost everyone relates to when they read the word "service".

When you write "Operator backed service", add a link to the term Operator. Use https://github.com/operator-framework.

  > A service created with odo creates multiple Kubernetes resources to establish a proper linking between the source component and target microservice, or 
* Devfile
  A portable YAML file responsible for your entire reproducible development environment. See [Devfile](../architecture/devfile.md) to know more about devfile.
  > A devfile can be considered similar to a manifest file in Kubernetes, but there is no equivalent of it in Kubernetes.
  

### Component vs. Application

> A component is a microservice, while an application is a group of microservices. To make a microservice belong to an application, all the resources belonging to the microservice are assigned a label.

Reviewer 2: ```suggestion
A component is akin to a microservice, while an application is a group of components.
```
> To make a microservice belong to an application, all the resources belonging to the microservice are assigned a label.

Removed this as it is more about how it's done under the hood, and we don't need to expose the implementation detail to the user/reader.

Author: @dharmit Do we need this section? I think that we covered the differences in the concepts section and I cannot think of anything new to put in the component vs application section.

Reviewer 2: We can remove it from here.

### odo services vs. Kubernetes services

> odo service helps in connecting one microservice to another, while a Kubernetes service helps in making a microservice accessible outside the cluster.

Reviewer 2: NO! It's the link that helps in connecting stuff. 

odo service is a service created from an Operator. k8s service definition looks good.

Reviewer: A kubernetes service is mainly an abstraction that serves to expose a set of Pods **in a stable way** (the service will keep the same IP during its life, when Pods can be scaled and restarted, each instance having a new and separate IP).

Services can expose the microservice at different levels depending on its type:
- clusterIP type on the cluster for inter-microservices communication, 
- node type to access from outside the Kubernetes cluster through the nodes
- loadBalancerType to access from outside the provider cluster, through some load balancer

Author: :grimacing: 

>How would you define an operator-backed service? If I use the term, I will still need to explain what a service is. How do I define a service?

Quoting the same comment as here https://github.com/openshift/odo/pull/4914#discussion_r670525665

Reviewer: A service in the odo sense is some external application that the user's application can interact with to gain some functionalities.

This wikipedia page can help find more inspiration: https://en.wikipedia.org/wiki/Service-oriented_architecture

Reviewer 2: We can get rid of the "odo service vs. k8s service" from this page as long as we explain odo services well. The comparison can be kept for Command Reference section.

> Creating an odo service will create a set of resources to establish a successful connection between the two microservices.

Reviewer 2: This is totally wrong. It's a link that helps setup connectivity between microservices. We don't need to have this under odo service vs k8s service.

Author: >This is totally wrong. It's a link that helps setup connectivity between microservices

Right, my bad. I mixed up service with a link while writing this.
